### PR TITLE
Solar System done

### PR DESCRIPTION
**Pull request by sarashahbaig**
# Solar System
Congratulations! You're submitting your assignment.

## Comprehension Questions

|  Question 	|  Answer
|---	|---	|
| When does the `initialize` method run? What does it do? | The `initalize` method runs when an instance of a class is created. That is, when we use the `.new` with a class to create an instance. The job of the `initialize` method is to create an instance of a class. 
| Why do you imagine we made our instance variables readable but not writable? | Because we want instance variables not to be modified by the user.
| How would your program be different if each planet was stored as a `Hash` instead of an instance of a class? | The program will have more code as we would write the keys for each hash. Secondly, we would not be able to do any validation, such for the mass not being a negative value. Third, we would not have custom methods to call on each single hash.
| How would your program be different if your `SolarSystem` class used a `Hash` instead of an `Array` to store the list of planets? | We would have to provide keys for each single planet. Secondly, we would not be able to call methods on those planets as we had defined in the `Planet` class.
| There is a software design principle called the SRP. The Single Responsibility Principle (SRP) says that each class should be responsible for exactly one thing. Do your classes follow SRP? What responsibilities do they have? | Yes, these two classes follow the SRP software design principle. The `Planet` class is responsible for how to create an instance of the same class and then also responsible for providing summary details for the same instance. The `SolarSystem` class is responsible for how an instance of the same class is created along with other functionality that belong to a solar system.
| How did you organize your `require` statements? Which files needed `require`s, and which did not? What is the pattern? | I used `require_relative` to include the `planet` and `solar system` files in the `main.rb` file. Only `main.rb` need `require`. It appears that `require` is used to include class files whenever we want to create an instance of those classes.

### CODE REVIEW

**Comment by tildeee** [comment]
# Solar System
## What We're Looking For

Feature | Feedback
--- |---
**Baseline** | 
Whitespace (indentation, vertical space, etc) | x
Variable names | x
Git hygiene | x, I would love to see more frequent, smaller commits on bigger projects in the future! I love seeing many, small commits that capture one small piece of code change
**Planet**  |
`initialize` method stores parameters as instance variables with appropriate reader methods |
`summary` method returns a string | x
**SolarSystem** |
`initialize` creates an empty list of planets | x
`add_planet` takes an instance of `Planet` and adds it to the list | x
`list_planets` returns a string | x
`find_planet_by_name` returns the correct instance of `Planet` | x
**CLI** |
Can list planets and quit | x
Can show planet details | x
Can add a planet | x
Complex functionality is broken out into separate methods | x
**Overall** |



Well done on this project, Sara! You did a great job on this project! Your code that practices object-oriented principles looks fantastic and exactly what we were looking for.


I want to highlight that your unit tests are wonderful and look great-- great work on that optional requirement!


However, there ends up being a small bug in your program-- particularly one around finding planets. I've added comments about it


That being said, good work on this project; keep it up!

File: lib/solar_system.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    end
+
+    def find_planet_by_name(planet_name)
+        # return  
+        @planets.each do |planet|           
+            if planet_name.downcase == planet.name.downcase
+                return planet
+            end
+        end
+        return "No planet with the #{planet_name} in this solar system!"
```
Comment (tildeee) [inline lib/solar_system.rb:29]: What happens when the planet wasn't found? We return this string. What happens in `main` when we get back this String after we call `.find_planet_by_name`?

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      elsif user_input == 2
+        planet_details(solar_system)
+      elsif user_input == 3
+        add_planet(solar_system)
+
+      end
+
+    end
+     
+  end
```
Comment (tildeee) [inline main.rb:49]: Your indentation gets off somewhere, so that this `end` that closes the `main` method is indented wrong. Keep an eye on correct indentation!

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+
+  def planet_details(solar_system)
+    print "Enter the name of the planet you want to learn about: "
+    planet_name = gets.chomp
+    planet = solar_system.find_planet_by_name(planet_name)
+    puts planet.summary
```
Comment (tildeee) [inline main.rb:56]: If the program asks me `"Enter the name of the planet you want to learn about: "` and I put in `9banana`, then we run `planet = solar_system.find_planet_by_name(planet_name)`... in this case, then `planet` has the value of the string `"No planet with the 9banana in this solar system!"` (which is what the method returns if no planet was found). Then, in this above line, we call `.summary` on `planet`... Which then breaks the program with a `NoMethodError`.

Ideally, we would change the `main` method so that the line `planet.summary` wouldn't run if we didn't find a matching `Planet`

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  def add_planet(solar_system)
+    print "Please enter the planet details: "
+    print "Enter planet name: "
+    planet_name = gets.chomp
+    print "Enter planet color: "
+    planet_color = gets.chomp
+    print "Enter planet mass: "
+    planet_mass = gets.chomp.to_f
```
Comment (tildeee) [inline main.rb:66]: Great job on doing the optional enhancement by letting `Planet`'s `initialize` method throw an error if the mass is invalid! That `raise` should remain in `Planet`. Also, you tested it in your unit tests, so that helps us understand as programmers that it's possible to reach that error.

However, when it comes to this `main` method, we don't want that error to ever get to the user. Because of that, I may recommend in the future ensuring that the user only inputs valid numbers here by putting this line in a `while` loop.

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    print "Enter planet distance from sun: "
+    planet_distance = gets.chomp.to_f
+    print "Enter a fun fact about this planet: "
+    planet_fun_fact = gets.chomp
+    
+    planet = Planet.new(planet_name,planet_color,planet_mass,planet_distance,planet_fun_fact)
```
Comment (tildeee) [inline main.rb:72]: Even though this line is really long, it's probably more readable to include a space after each comma. Another way to break up this line in Ruby is to use this syntax:

```ruby
  planet = Planet.new(planet_name,
    planet_color,
    planet_name,
    planet_distance,
    planet_fun_fact
  )
```

File: test/planet_test.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
+require 'minitest/autorun'
+require 'minitest/reporters'
+require 'minitest/skip_dsl'
+
+require_relative '../lib/planet'
+
+# 
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe 'planet' do
```
Comment (tildeee) [inline test/planet_test.rb:10]: Because all of these `it` block tests share the same `describe 'planet'` description, it may be okay to put all of these `it` blocks as next to each other inside of _one_ `describe 'planet'`

File: test/planet_test.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    it 'check if color and fun_fact is a string' do
+        earth = Planet.new('Earth', 'green', 5.972e24, 5.5, 'Only planet known to support life')
+        expect(earth.fun_fact).must_be_instance_of String
+        expect(earth.color).must_be_instance_of String
+    end
+    it 'raises an ArgumentError if number is less then zero' do
+    
+        expect{Planet.new('Earth', 'green', -100, 5.5, 'Only planet known to support life')}.must_raise ArgumentError
```
Comment (tildeee) [inline test/planet_test.rb:42]: Well done on writing this edge case test! :) This is perfect!
